darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+
+from darshan.report import *
+
+import matplotlib
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1 import make_axes_locatable 
+import numpy as np
+
+def plot_dxt_heatmap(report, 
+        xbins=10, ybins=None, 
+        group_by="rank", 
+        mods=None, ops=None, 
+        display_values=False, cmap=None, figsize=None, 
+        amplify=False):
+    """
+    Generates a heatmap plot from a report with DXT traces.
+
+    Args:
+        report (darshan.DarshanReport): report to generate plot from
+        xbins (int): number of bins on the x axis
+        ybins (int): number of bins on the y axis
+        group_by (str): attribute to group by (e.g., rank, hostname)
+        mods (list): modules to include in heatmap (e.g., ['DXT_POSIX', 'DXT_MPIIO'])
+        ops (list): operations to consider (e.g., ['read', 'write']
+        display_values (bool): show values per heatmap field
+        cmap: overwrite colormap (see matplotlib colormaps)
+        figsize: change figure size (see matplotlib figsize)
+
+        amplify (int): paint neighbouring cells e.g., when working with many ranks
+    """
+
+    runtime = report.end_time - report.start_time       # timedelta
+    runtime = runtime.seconds + runtime.microseconds    # to float/int
+    if runtime == 0:
+        runtime = 1
+
+        
+    if mods is None:
+        mods = ['DXT_POSIX', 'DXT_MPIIO']
+    
+    if ops is None:
+        ops = ['read', 'write']
+    
+    
+    def get_hostname_mapper(report):
+        """ Determine which ranks map to which hostname. """
+        hostnames = {}
+        i = 0
+        for mod in mods:
+            for rec in report.records[mod]:
+                hostname = rec['hostname']
+                if hostname not in hostnames:
+                    hostnames[hostname] = i
+                    i += 1
+        return hostnames
+   
+
+    hostnames = get_hostname_mapper(report)
+    ranks = report.metadata['job']['nprocs']
+    
+    # set ybins, yticks and ylabel to match choice of group_by 
+    if group_by in ['hostname', 'node']:
+        ybins = len(hostnames)
+        yticks = list(hostnames.keys())
+        ylabel = "Hostname"
+    else:
+        if ybins is None:
+            ybins = ranks 
+            ylabel = "Rank"
+            #yticks = [str(x) for x in range(ybins)] 
+        else:
+            ylabel = f"Binned Ranks (binsize={int(ranks/ybins)} ranks)"
+
+    
+    # heatmap to be populated with event counts
+    events = np.zeros((ybins, xbins))    
+    
+    for mod in mods:
+        for op in ops:
+            for rec in report.records[mod]:
+                
+                if group_by in ['hostname', 'node']:
+                    ybin = hostnames[rec['hostname']]
+                else:
+                    ybin = int((rec['rank'] /ranks) * ybins)
```

Comment:
For log files with a few thousand ranks you could save a few thousand unnecessary multiplications with this as far as I can tell. I have few more thoughts on this nested for loop I'll add in a separate comment.

```diff
--- a/darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py
+++ b/darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py
@@ -76,14 +76,15 @@ def plot_dxt_heatmap(report,
     # heatmap to be populated with event counts
     events = np.zeros((ybins, xbins))    
     
+    if group_by == "rank":
+        ybin_denom = ranks / ybins
     for mod in mods:
         for op in ops:
             for rec in report.records[mod]:
-                
                 if group_by in ['hostname', 'node']:
                     ybin = hostnames[rec['hostname']]
                 else:
-                    ybin = int((rec['rank'] /ranks) * ybins)
+                    ybin = int(rec['rank'] / ybin_denom)

```

Reply:
Given that there is the other DXT heatmap I marked this as resolved :)